npeps/npep-187-ports-and-protocols.md

```diff
@@ -0,0 +1,222 @@
+# NPEP-187: More protocols support
+
+* Issue: [#187](https://github.com/kubernetes-sigs/network-policy-api/issues/187)
+* Status: Provisional
+
+## TLDR
+
+Change existing `ports` selector to allow more protocols in the future.
+
+## Goals
+
+Change existing `ports` selector to allow more protocols in the future.
+
+## Non-Goals
+
+Adding new protocols to the API.
+
+## Introduction
+
+Currently only TCP, UDP, and SCTP protocols are supported, but there are more protocols (ICMP, ICMPv6 are the most popular requests) 
+that may be useful in the future, so we want to leave space for future extensions.
+
+## User-Stories/Use-Cases
+
+Story 1: Protocols extension
+
+As a cluster admin I want to be able to use protocols other than TCP, UDP, and SCTP in my CNP.
+For example, I want to only allow ICMP connections to implement health monitoring and deny everything else.
+
+## API
+
+As we only support port-based protocols (TCP, UDP, and SCTP), current protocol matching is expressed as
+```yaml
+ports:
+  - portNumber:
+      protocol: TCP
+      port: 80
+  - portRange:
+      protocol: UDP
+      start: 1
+      end: 100
+```
+To enable more protocols support in the future, we need to change the high-level field name from `port` to `protocol` to be more generic.
+
+In the `protocols` filter design we decided to start from the YAML that we want (i.e. YAML that is easy to write and understand).
+We want to make popular filters (TCP and UDN ports) **easy** and less popular filters (like ICMP and named ports) **possible**.
+There are multiple options to support this:
+
+### Option 1: Protocols list with protocol discriminator
+
+```yaml
+protocols:
+  - protocol: TCP
+    port: 
+      number: 8080
+  - protocol: UDP
+    port:
+      range: 
+        start: 8080
+        end: 9090 
+  - namedPort: http
+    # protocol: "" # must be empty string: defaulted / omitempty
+```
+
+with the validation like so:
+- empty `protocols` list is not allowed
+- at least 1 field for each `protocol` element must be set
+- if `namedPort` is set, `protocol` must be unset
+- if `protocol` is TCP/UDP/SCTP, `port` must be set
+
+### Option2: Join by protocol
+
+```yaml
+protocols:
+  tcp:
+    - port:
+        number: 8080
+    - port:
+        range: 
+          start: 8080
+          end: 9090
+  udp:
+    - port:
+        number: 8080
+    - port:
+        number: 9090
+  namedPorts:
+    - http
+    - monitoring
+  icmp:
+    - type: 7
+      code: 3
+```
+
+In this case it makes more sense to also flatten the `port` match, because for TCP/UDP/SCTP protocols, port is the only option and is always required:
+
+```yaml
+protocols:
+  tcp:
+    - portNumber: 8080
```

Comment:
```suggestion
    - port: 8080
```
We don't say `portNumberRange`, could we shorten here too?

Reply:
huh makes sense